### ISSUE TITLE
[SLPVectorizer] Crash during horizontal reduction (size() >= N && "Dropping more elements than exist")

### ISSUE THREAD

**Issue by annamthomas** [llvm/llvm-project#64568]
We crash in SLPVectorizer during horizontal reduction stating we remove more elements than present in the ArrayRef.

Here's the reproducer.
`cat upstream.ll`
```ll
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128-ni:1-p2:32:8:8:32-ni:2"
target triple = "x86_64-unknown-linux-gnu"

define void @ham.27() #0 {
bb:
  br label %bb169

bb169:                                            ; preds = %bb
  %shl183 = shl i32 0, 6
  %shl187 = shl i32 0, 8
  %or188 = or i32 %shl183, 1
  %add189 = add i32 %or188, %shl183
  %add190 = add i32 0, %add189
  %shl191 = shl i32 0, 10
  %or192 = or i32 %shl187, 1
  %add193 = add i32 %or192, %shl187
  %add194 = add i32 %add190, %add193
  %or196 = or i32 %shl191, 1
  %add197 = add i32 %or196, %shl191
  %add198 = add i32 %add194, %add197
  %shl199 = shl i32 0, 14
  %add202 = add i32 %add198, poison
  %or203 = or i32 %shl199, 1
  %add204 = add i32 %or203, %shl199
  %add205 = add i32 %add202, %add204
  %sub206 = sub i32 poison, %add205
  store atomic i32 %sub206, ptr addrspace(1) poison unordered, align 4
  unreachable
}
```

Reproduce with
```shell
opt -passes=slp-vectorizer upstream.ll -debug -mattr=+avx
```

Failure is `Assertion failed: (size() >= N && "Dropping more elements than exist")`, function `drop_back`, file `ArrayRef.h`, line 211.

**Comment by annamthomas** [comment]
Output from debug:
```
SLP: Analyzing blocks in ham.27.
SLP:  bundle:   %or203 = or i32 %shl199, 1
SLP:  initialize schedule region to   %or203 = or i32 %shl199, 1
SLP:  extend schedule region start to   %or196 = or i32 %shl191, 1
SLP:  extend schedule region start to   %or192 = or i32 %shl187, 1
SLP:  extend schedule region start to   %or188 = or i32 %shl183, 1
SLP: try schedule bundle [  %or203 = or i32 %shl199, 1;  %or196 = or i32 %shl191, 1;  %or192 = or i32 %shl187, 1;  %or188 = or i32 %shl183, 1] in block bb169
SLP:       update deps of [  %or203 = or i32 %shl199, 1;  %or196 = or i32 %shl191, 1;  %or192 = or i32 %shl187, 1;  %or188 = or i32 %shl183, 1]
SLP:       update deps of /   %or196 = or i32 %shl191, 1
SLP:       update deps of /   %or192 = or i32 %shl187, 1
SLP:       update deps of /   %or188 = or i32 %shl183, 1
SLP:       update deps of   %add189 = add i32 %or188, %shl183
SLP:       update deps of   %add190 = add i32 0, %add189
SLP:       update deps of   %add194 = add i32 %add190, %add193
SLP:       update deps of   %add198 = add i32 %add194, %add197
SLP:       update deps of   %add202 = add i32 %add198, poison
SLP:     gets ready on update:   %add202 = add i32 %add198, poison
SLP:       update deps of   %add193 = add i32 %or192, %shl187
SLP:       update deps of   %add197 = add i32 %or196, %shl191
SLP:    initially in ready list:   %add202 = add i32 %add198, poison
SLP:   schedule   %add202 = add i32 %add198, poison
SLP:    gets ready (def):   %add198 = add i32 %add194, %add197
SLP:   schedule   %add198 = add i32 %add194, %add197
SLP:    gets ready (def):   %add194 = add i32 %add190, %add193
SLP:    gets ready (def):   %add197 = add i32 %or196, %shl191
SLP:   schedule   %add197 = add i32 %or196, %shl191
SLP:   schedule   %add194 = add i32 %add190, %add193
SLP:    gets ready (def):   %add190 = add i32 0, %add189
SLP:    gets ready (def):   %add193 = add i32 %or192, %shl187
SLP:   schedule   %add193 = add i32 %or192, %shl187
SLP:   schedule   %add190 = add i32 0, %add189
SLP:    gets ready (def):   %add189 = add i32 %or188, %shl183
SLP:   schedule   %add189 = add i32 %or188, %shl183
SLP:    gets ready (def): [  %or203 = or i32 %shl199, 1;  %or196 = or i32 %shl191, 1;  %or192 = or i32 %shl187, 1;  %or188 = or i32 %shl183, 1]
SLP: We are able to schedule this bundle.
SLP: added a vector of un/bin op.
SLP: We are able to schedule this bundle.
SLP: added a vector of un/bin op.
SLP: Gathering due to C,S,B,O, small shuffle. 
SLP: Gathering due to C,S,B,O, small shuffle. 
SLP: Gathering due to C,S,B,O, small shuffle. 
SLP: Checking user:  %add204 = add i32 %or203, %shl199.
SLP: Checking user:  %add197 = add i32 %or196, %shl191.
SLP: Checking user:  %add193 = add i32 %or192, %shl187.
SLP: Checking user:  %add189 = add i32 %or188, %shl183.
SLP: Need to extract: Extra arg from lane 0 from   %shl199 = shl i32 0, 14.
SLP: Checking user:  %add204 = add i32 %or203, %shl199.
SLP: Checking user:  %or203 = or i32 %shl199, 1.
SLP: 	Internal user will be removed:  %or203 = or i32 %shl199, 1.
SLP: Need to extract: Extra arg from lane 1 from   %shl191 = shl i32 0, 10.
SLP: Checking user:  %add197 = add i32 %or196, %shl191.
SLP: Checking user:  %or196 = or i32 %shl191, 1.
SLP: 	Internal user will be removed:  %or196 = or i32 %shl191, 1.
SLP: Need to extract: Extra arg from lane 2 from   %shl187 = shl i32 0, 8.
SLP: Checking user:  %add193 = add i32 %or192, %shl187.
SLP: Checking user:  %or192 = or i32 %shl187, 1.
SLP: 	Internal user will be removed:  %or192 = or i32 %shl187, 1.
SLP: Need to extract: Extra arg from lane 3 from   %shl183 = shl i32 0, 6.
SLP: Checking user:  %add189 = add i32 %or188, %shl183.
SLP: Checking user:  %or188 = or i32 %shl183, 1.
SLP: 	Internal user will be removed:  %or188 = or i32 %shl183, 1.
SLP: Calculating cost for tree of size 5.
SLP: Calculated costs for Tree:
0.
Operand 0:
    %shl199 = shl i32 0, 14
    %shl191 = shl i32 0, 10
    %shl187 = shl i32 0, 8
    %shl183 = shl i32 0, 6
Operand 1:
  i32 1
  i32 1
  i32 1
  i32 1
Scalars: 
    %or203 = or i32 %shl199, 1
    %or196 = or i32 %shl191, 1
    %or192 = or i32 %shl187, 1
    %or188 = or i32 %shl183, 1
State: Vectorize
MainOp:   %or203 = or i32 %shl199, 1
AltOp:   %or203 = or i32 %shl199, 1
VectorizedValue: NULL
ReuseShuffleIndices: Empty
ReorderIndices: 
UserTreeIndices: 
SLP: Costs:
SLP:     ReuseShuffleCost = 0
SLP:     VectorCost = 1
SLP:     ScalarCost = 4
SLP:     ReuseShuffleCost + VecCost - ScalarCost = -3
SLP: Adding cost -3 for bundle that starts with   %or203 = or i32 %shl199, 1.
SLP: Current total cost = -3
SLP: Calculated costs for Tree:
1.
Operand 0:
  i32 0
  i32 0
  i32 0
  i32 0
Operand 1:
  i32 14
  i32 10
  i32 8
  i32 6
Scalars: 
    %shl199 = shl i32 0, 14
    %shl191 = shl i32 0, 10
    %shl187 = shl i32 0, 8
    %shl183 = shl i32 0, 6
State: Vectorize
MainOp:   %shl199 = shl i32 0, 14
AltOp:   %shl199 = shl i32 0, 14
VectorizedValue: NULL
ReuseShuffleIndices: Empty
ReorderIndices: 
UserTreeIndices: {User:0 EdgeIdx:0}, 
SLP: Costs:
SLP:     ReuseShuffleCost = 0
SLP:     VectorCost = 2
SLP:     ScalarCost = 4
SLP:     ReuseShuffleCost + VecCost - ScalarCost = -2
SLP: Adding cost -2 for bundle that starts with   %shl199 = shl i32 0, 14.
SLP: Current total cost = -5
SLP: Adding cost 0 for bundle that starts with i32 0.
SLP: Current total cost = -5
SLP: Adding cost 0 for bundle that starts with i32 14.
SLP: Current total cost = -5
SLP: Adding cost 0 for bundle that starts with i32 1.
SLP: Current total cost = -5
SLP: #LV: 1 shl199, Looking at   %shl199 = shl i32 0, 14
SLP: Spill Cost = 0.
SLP: Extract Cost = 4.
SLP: Total Cost = -1.
SLP: Adding cost 0 for reduction that starts with   %or203 = or i32 %shl199, 1 (It is a splitting reduction)
SLP: Found cost = -1 for reduction
SLP: Vectorizing horizontal reduction at cost:-1. (HorRdx)
SLP: schedule block bb169
SLP:    initially in ready list:   %add202 = add i32 %add198, poison
SLP:   schedule   %add202 = add i32 %add198, poison
SLP:    gets ready (def):   %add198 = add i32 %add194, %add197
SLP:   schedule   %add198 = add i32 %add194, %add197
SLP:    gets ready (def):   %add194 = add i32 %add190, %add193
SLP:    gets ready (def):   %add197 = add i32 %or196, %shl191
SLP:   schedule   %add197 = add i32 %or196, %shl191
SLP:   schedule   %add194 = add i32 %add190, %add193
SLP:    gets ready (def):   %add190 = add i32 0, %add189
SLP:    gets ready (def):   %add193 = add i32 %or192, %shl187
SLP:   schedule   %add193 = add i32 %or192, %shl187
SLP:   schedule   %add190 = add i32 0, %add189
SLP:    gets ready (def):   %add189 = add i32 %or188, %shl183
SLP:   schedule   %add189 = add i32 %or188, %shl183
SLP:    gets ready (def): [  %or203 = or i32 %shl199, 1;  %or196 = or i32 %shl191, 1;  %or192 = or i32 %shl187, 1;  %or188 = or i32 %shl183, 1]
SLP:   schedule [  %or203 = or i32 %shl199, 1;  %or196 = or i32 %shl191, 1;  %or192 = or i32 %shl187, 1;  %or188 = or i32 %shl183, 1]
SLP: Extracting 4 values .
SLP: 	validating user:  %add204 = add i32 %or203, 0.
SLP: 	Erasing scalar:  %or203 = or i32 0, 1.
SLP: 	validating user:  %add197 = add i32 %or196, 0.
SLP: 	Erasing scalar:  %or196 = or i32 0, 1.
SLP: 	validating user:  %add193 = add i32 %or192, 0.
SLP: 	Erasing scalar:  %or192 = or i32 0, 1.
SLP: 	validating user:  %add189 = add i32 %or188, 0.
SLP: 	Erasing scalar:  %or188 = or i32 0, 1.
SLP: 	Erasing scalar:  %shl199 = shl i32 0, 14.
SLP: 	Erasing scalar:  %shl191 = shl i32 0, 10.
SLP: 	Erasing scalar:  %shl187 = shl i32 0, 8.
SLP: 	Erasing scalar:  %shl183 = shl i32 0, 6.
Assertion failed: (size() >= N && "Dropping more elements than exist"), function drop_back, file ArrayRef.h, line 211.
```

**Comment by annamthomas** [comment]
Adding a couple more debug output, I can see why the error happens, but not the root cause.
We are dropping elements from the `ReducedValsToOps` array entry for RdxVal = i32 0.
```
RdxVal: i32 0  NumOps: 4
ReducedValsToOps Array  for RdxVal
  %add190 = add i32 0, %add189
```
As we can see, we have only one entry for this `i32 0`. We're trying to remove 4. It could be that NumOps is incorrect, which is calculated as `unsigned NumOps = VectorizedVals.lookup(RdxVal);`. 

@alexey-bataev  Could you pls take a look?

**Comment by alexey-bataev** [comment]
Fixed in 22165071711a3b4506487212fe7e0fa1fb0ba6a5
